bokehjs/src/coffee/core/selection_manager.ts

```diff
@@ -10,47 +10,152 @@ import {DataSource} from "models/sources/data_source"
 export type Renderer = any
 export type RendererView = any
 
+export type HitTestResult = Selection | null
+
+export abstract class SelectionPolicy extends Model {
+
+  hit_test(geometry: Geometry, renderer_views: RendererView[]): HitTestResult{
+    return null
+  }
+
+  do_selection(hit_test_result: HitTestResult | null, renderer_views: RendererView[], final: boolean, append: boolean): boolean {
+    if (hit_test_result === null) {
+      return false
+    } else {
+      const source = renderer_views[0].model.data_source
+      source.selected.update(hit_test_result, final, append)
+      source.select.emit()
+      return !source.selected.is_empty()
+    }
+  }
+}
+
+SelectionPolicy.prototype.type = "SelectionPolicy"
+
+export class IntersectRenderers extends SelectionPolicy {
+
+  hit_test(geometry: Geometry, renderer_views: RendererView[]): HitTestResult {
+    let hit_test_result_renderers = []
+    for (let r of renderer_views) {
+      let result = r.hit_test(geometry)
+      if (result !== null)
+        hit_test_result_renderers.push(result)
+    }
+    if (hit_test_result_renderers.length > 0) {
+      const hit_test_result = hit_test_result_renderers[0]
+      for (let hit_test_result_other of hit_test_result_renderers) {
+        hit_test_result.update_through_intersection(hit_test_result_other)
+      }
+      return hit_test_result
+    } else {
+      return null
+    }
+  }
+}
+
+IntersectRenderers.prototype.type = "IntersectRenderers"
+
+export class UnionRenderers extends SelectionPolicy {
+
+  hit_test(geometry: Geometry, renderer_views: RendererView[]): HitTestResult {
+    let hit_test_result_renderers = []
+    for (let r of renderer_views) {
+      let result = r.hit_test(geometry)
+      if (result !== null)
+        hit_test_result_renderers.push(result)
+    }
+    if (hit_test_result_renderers.length > 0) {
+      const hit_test_result = hit_test_result_renderers[0]
+      for (let hit_test_result_other of hit_test_result_renderers) {
+        hit_test_result.update_through_union(hit_test_result_other)
+      }
+      return hit_test_result
+    } else {
+      return null
+    }
+  }
+}
+
+UnionRenderers.prototype.type = "UnionRenderers"
+
 export class SelectionManager extends HasProps {
 
+  selection_policy: SelectionPolicy
   source: DataSource
-
-  selector: Selector
-  inspectors: {[key: string]: HasProps}
+  inspectors: {[key: string]: Selection}
 
   initialize(attrs: any, options: any): void {
     super.initialize(attrs, options)
-    this.selector = new Selector()
     this.inspectors = {}
   }
 
   select(renderer_views: RendererView[], geometry: Geometry, final: boolean, append: boolean = false): boolean {
+    // divide renderers into glyph_renderers or graph_renderers
+    const glyph_renderer_views: RendererView[] = []
+    const graph_renderer_views: RendererView[] = []
+    for (let r of renderer_views) {
+      if (r.model.type == 'GlyphRenderer'){
```

Comment:
Is there a reason to use `type`? The right solution is to use `instanceof`, because this way you respect the fact that models can be extended. `type` should be used only for serialization purpose and extremely rare occasions when nothing else works (e.g. mutually recursive imports that result in infinite recursion).

Reply:
I just tried this again, and think I may have originally tried it and then run into this issue where ColumnarDataSource imports SelectionManager which imports GlyphRenderer which imports RemoteDataSource which breaks. This is maybe because RemoteDataSource inherits from ColumnarDataSource?